NAS-109762 / 21.04 / prevent the ability to share raw disks via iscsi on SCALE

yocalebo opened the pull request:
- defeats the purpose of our product to allow the sharing of raw disks (we expect ZFS to be used underneath the hood)
- SCST has a build time warning about data corruption in certain circumstances. If a raw disk is shared via SCST and if the `sd`, `sg` or `st` drivers are used for accessing/modifying disk state then it can lead to data corruption. On SCALE Enterprise, we use the `sd` and `sg` drivers.

sonicaj reviewed (commented):
I have following concerns:
1) What will happen to users who have already configured disks ?
2) We have more disks logic in extent service, can we remove that as well please because we are not using it anymore ?

william-gr commented on src/middlewared/middlewared/alembic/versions/12.1/2021-03-15_14-22_remove_rawdisks_from_iscsi.py:
```diff
@@ -0,0 +1,48 @@
+            f'SELECT id, iscsi_target_id FROM {tartoext_table} WHERE iscsi_extent_id == {disk_id}'
+        ).fetchall():
+            tar_ids.add(tar_id)
+            conn.execute(f'DELETE FROM {tartoext_table} WHERE id == {id}')
```
Hmm I never saw SQL queries with `==`

yocalebo replied:
> Hmm I never saw SQL queries with `==`

sqlite> SELECT * FROM services_iscsitargetextent where id == 2;
                                 id = 2
           iscsi_target_extent_name = disk01
         iscsi_target_extent_serial = 5624ba25976f000
           iscsi_target_extent_type = Disk
           iscsi_target_extent_path = {serial}drive-scsi1
       iscsi_target_extent_filesize = 0
      iscsi_target_extent_blocksize = 512
     iscsi_target_extent_pblocksize = 0
iscsi_target_extent_avail_threshold =
        iscsi_target_extent_comment = disk01
            iscsi_target_extent_naa = 0x6589cfc000000857b205f610a35fb3bd
   iscsi_target_extent_insecure_tpc = 1
            iscsi_target_extent_xen = 0
            iscsi_target_extent_rpm = SSD
             iscsi_target_extent_ro = 0
        iscsi_target_extent_enabled = 1
         iscsi_target_extent_vendor = TrueNAS

themylogin replied:
I did not know that works either. We don't use `==` in SQL queries

yocalebo replied:
So are you all asking me to change to `=`?

william-gr replied:
Yes, `==` is not SQL standard as far as I can understand, maybe only works in sqlite3, its not consistent

```
postgres=# select 1=1;
 ?column? 
----------
 t
(1 row)

postgres=# select 1==1;
ERROR:  operator does not exist: integer == integer
LINE 1: select 1==1;
                ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

sonicaj commented on src/middlewared/middlewared/plugins/iscsi.py:
```diff
@@ -800,7 +798,8 @@ async def clean_type_and_path(self, data, schema_name, verrors):
 
-        if extent_type == 'Disk':
+        # raw disks being shared via iscsi is only supported on freeBSD
+        if osc.IS_FREEBSD and extent_type == 'Disk':
```
How would this happen even on FreeBSD ?

yocalebo replied:
Not sure I understand. This would happen if we release a core image based off master repo?

william-gr replied:
You just removed all Disk from database so this would never be true and its easier/better to remove the code too?

themylogin commented on src/middlewared/middlewared/plugins/iscsi.py:
```diff
@@ -819,10 +818,10 @@ async def clean_type_and_path(self, data, schema_name, verrors):
                 raise verrors  # They need this for anything else
 
             if '/iocage' in path:
```
@william-gr can we start removing FreeBSD stuff?

william-gr commented on src/middlewared/middlewared/plugins/iscsi.py:
```diff
@@ -910,8 +852,7 @@ async def disk_choices(self, exclude):
         for e in exclude:
-            if e:
-                zvol_query_filters.append(('path', '!=', e))
+            zvol_query_filters.append(('path', '!=', e))
```
Since we are changing this, could probable be
`zvol_query_filters.append(('path', 'nin', exclude))`

thoughts?

william-gr reviewed (commented):
Shouldn't we still remove the old DISK entries from database?

yocalebo commented:
> Shouldn't we still remove the old DISK entries from database?

@william-gr yeah, when I pushed rebased PR I forgot to include migration file 😁  I've fixed. Also, I've updated the docstring for the `disk_choices` method. I really don't like the whole `ignore` parameter and as best as I can tell nothing in middleware (or webui) is using that so I'm inclined to remove it.
